src/cdk/table/coalesced-style-scheduler.ts

```diff
@@ -35,7 +42,46 @@ export class _CoalescedStyleScheduler implements OnDestroy {
   private _currentSchedule: _Schedule | null = null;
   private readonly _destroyed = new Subject<void>();
 
-  constructor(private readonly _ngZone: NgZone) {}
+  private readonly _earlyReadTasks: (() => unknown)[] = [];
+  private readonly _writeTasks: (() => unknown)[] = [];
+  private readonly _readTasks: (() => unknown)[] = [];
+
+  constructor(private readonly _ngZone: NgZone) {
```

Comment:
AFAIK `afterRender` has its own internal coalescing mechanism. Couldn't we switch to using `afterRender` directly in the table and remove the  `_CoalescedStyleScheduler` altogether?

Reply:
Not really possible alas as afterRender has a few limitations.

1) It can only be invoked during injection context, wheras I need to have things happen at arbitrary times.
2) afterRender has no mechanism to flush outside of NgZone/CD, which is needed sometimes in the table.
3) I tried to move columnResize off of the old _CoalescedStyleScheduler hooks, but could not get that to work, so is a blocker for now at least.

Reply:
applicationRef.tick has additional side effects performance-wise. Given there are no changes to application state, just CSS, that does not seem like the best choice here.

Also, question on afterRender() - can you call safely afterRender/afterNextRender during afterRender? There are cases where it needs to do that.

Reply:
If/when you get to trying to submit this change again, I'd love to discuss this more. `afterRender` is still in developer preview and its behavior is being tuned to ensure it satisfies required use-cases.

> afterRender has no mechanism to flush outside of NgZone/CD, which is needed sometimes in the table.
...
 applicationRef.tick has additional side effects performance-wise. Given there are no changes to application state, just CSS, that does not seem like the best choice here.
...
 Given there are no changes to application state, just CSS, that does not seem like the best choice here.


We're working on several things that can help address this:

1. Guaranteeing that render hooks run, even when no change detection is scheduled: https://github.com/angular/angular/pull/54083/commits/80b953a92eca29f7b564b1d520b5c9ed3b6af7c2
2. Not refreshing views if we only need to flush `afterRender` hooks: https://github.com/angular/angular/pull/54083/commits/51dfdbb9de0eef44e21ab5f727f8225c7e43031d
3. Things that notify the scheduler (like `afterRender` hooks) should work when outside the zone (even in applications that use ZoneJS): https://github.com/angular/angular/issues/53844